django/db/backends/postgresql/compiler.py

```diff
@@ -43,7 +45,7 @@ def assemble_as_sql(self, fields, value_rows):
         db_types = [field.db_type(self.connection) for field in fields]
         # Abort if any of the fields are arrays as UNNEST indiscriminately
         # flatten them instead of reducing their nesting by one.
-        if any(db_type.endswith("[]") for db_type in db_types):
+        if any(re.search(r".+\[\d*\]$", db_type) for db_type in db_types):
```

Comment:
Looking at it we could avoid the `re` usage by doing

```suggestion
        if any(db_type.endswith("]") for db_type in db_types):
```

since [Postgres reserves the usage of brackets for arrays](https://www.postgresql.org/docs/current/arrays.html#ARRAYS-DECLARATION) so I'll push a commit that does that instead.